src/expr/dtype.cpp

```diff
@@ -0,0 +1,743 @@
+{
+  Assert(item.getType().isSelector());
+  return cindexOfInternal(item);
+}
+size_t DType::cindexOfInternal(Node item)
+{
+  if (item.getKind() == APPLY_TYPE_ASCRIPTION)
+  {
+    return cindexOf(item[0]);
+  }
+  Assert(item.hasAttribute(DTypeConsIndexAttr()));
+  return item.getAttribute(DTypeConsIndexAttr());
+}
+
+bool DType::resolve(const std::map<std::string, TypeNode>& resolutions,
+                    const std::vector<TypeNode>& placeholders,
+                    const std::vector<TypeNode>& replacements,
+                    const std::vector<TypeNode>& paramTypes,
+                    const std::vector<TypeNode>& paramReplacements)
+{
+  Trace("datatypes-init") << "DType::resolve: " << std::endl;
+  Assert(!d_resolved);
+  Assert(resolutions.find(d_name) != resolutions.end());
+  Assert(placeholders.size() == replacements.size());
+  Assert(paramTypes.size() == paramReplacements.size());
+  Assert(getNumConstructors() > 0);
+  TypeNode self = (*resolutions.find(d_name)).second;
+  Assert(&self.getDType() == this);
+  d_resolved = true;
+  size_t index = 0;
+  for (std::shared_ptr<DTypeConstructor> ctor : d_constructors)
+  {
+    Trace("datatypes-init") << "DType::resolve ctor " << std::endl;
+    if (!ctor->resolve(self,
+                       resolutions,
+                       placeholders,
+                       replacements,
+                       paramTypes,
+                       paramReplacements,
+                       index))
+    {
+      return false;
+    }
+    ctor->d_constructor.setAttribute(DTypeIndexAttr(), index);
+    ctor->d_tester.setAttribute(DTypeIndexAttr(), index++);
+    Assert(ctor->isResolved());
+    Trace("datatypes-init") << "DType::resolve ctor finished" << std::endl;
+  }
+  d_self = self;
+
+  d_involvesExt = false;
+  d_involvesUt = false;
+  for (const std::shared_ptr<DTypeConstructor> ctor : d_constructors)
+  {
+    if (ctor->involvesExternalType())
+    {
+      d_involvesExt = true;
+    }
+    if (ctor->involvesUninterpretedType())
+    {
+      d_involvesUt = true;
+    }
+  }
+
+  if (isSygus())
+  {
+    // all datatype constructors should be sygus and have sygus operators whose
+    // free variables are subsets of sygus bound var list.
+    std::unordered_set<Node, NodeHashFunction> svs;
+    for (const Node& sv : d_sygus_bvl)
+    {
+      svs.insert(sv);
+    }
+    for (unsigned i = 0, ncons = d_constructors.size(); i < ncons; i++)
+    {
+      Node sop = d_constructors[i]->getSygusOp();
+      Assert(!sop.isNull())
+          << "Sygus datatype contains a non-sygus constructor";
+      std::unordered_set<Node, NodeHashFunction> fvs;
+      expr::getFreeVariables(sop, fvs);
+      for (const Node& v : fvs)
+      {
+        if (svs.find(v) == svs.end())
+        {
+          // Print a warning and return false, indicating we should abort,
+          // since this datatype is not well formed.
+          Warning() << "Sygus constructor has an operator with a free variable "
```

Comment:
Shouldn't this be an error?

Reply:
The error should be propagated upwards to the user level so we can give a proper (pretty) exception.

Reply:
Hmm, so why print a warning at all in that case?